File: BinarySearchTree.cpp
#include <iostream>
using namespace std;

class BST {
  struct node {
    int data;
    node* left;
    node* right;
  };

  public:
    node* root;

    BST() {
      root = NULL;
    }

    ~BST() {
      root = deleteTree(root);
    }

    void insert(int x) {
      root = insert(x, root);
    }

    void remove(int x) {
      root = remove(x, root);
    }

    void inOrder() {
      cout << "In order: ";
      inOrder_(root);
      cout << endl << endl;
    }

    void preOrder() {
      cout << "Pre order: ";
      preOrder_(root);
      cout << endl << endl;
    }

    void postOrder() {
      cout << "Post order: ";
      postOrder_(root);
      cout << endl << endl;
    }

    void levelOrder() {
      cout << "Level order: ";
      int level = 1;
      while (levelOrder_(root, level)) level++;
      cout << endl << endl;
    }

    void search(int x) {
      root = find(root, x);
      cout << "Found: " << root -> data;
      cout << endl << endl;
    }

    void findMin() {
      root = findMin_(root);
      cout << "Found: " << root -> data;
      cout << endl << endl;
    }


  node* insert(int x, node* itr) {
    if (itr == NULL) {
      itr = new node;
      itr -> data = x;
      itr -> left = itr -> right = NULL;
    } else if(x < itr -> data) itr -> left = insert(x, itr -> left);
    else if(x > itr -> data) itr -> right = insert(x, itr -> right);
    return itr;
  }

  node* remove(int x, node* itr) {
    node* temp;
    if (itr == NULL)  return NULL;
    else if (x < itr -> data) itr -> left = remove(x, itr -> left);
    else if (x > itr -> data) itr -> right = remove(x, itr -> right);
    else if (itr -> left && itr -> right) {
      temp = findMin_(itr -> right);
      itr -> data = temp -> data;
      itr -> right = remove(itr -> data, itr -> right);
    } else {
      temp = itr;
      if (itr -> left == NULL) itr = itr -> right;
      else if (itr -> right == NULL) itr = itr -> left;
      delete temp;
    }

    return itr;
  }

  node* deleteTree(node* itr) {
    if (itr == NULL) return NULL;
    deleteTree(itr -> left);
    deleteTree(itr -> right);
    delete itr;
    return NULL;
  }

  void inOrder_(node* itr) {
    if (itr == NULL) return;
    inOrder_(itr -> left);
    cout << itr -> data << " ";
    inOrder_(itr -> right);
  }

  void preOrder_(node* itr) {
    if (itr == NULL) return;
    cout << itr -> data << " ";
    inOrder_(itr -> left);
    inOrder_(itr -> right);
  }

  void postOrder_(node* itr) {
    if (itr == NULL) return;
    inOrder_(itr -> left);
    inOrder_(itr -> right);
    cout << itr -> data << " ";
  }

  bool levelOrder_(node* root, int level) {
  	if (root == nullptr) return false;
  	if (level == 1) {
  		cout << root -> data << " ";
  		return true;
  	}

  	bool left = levelOrder_(root -> left, level - 1);
  	bool right = levelOrder_(root -> right, level - 1);

  	return left || right;
  }

  node* find(node* itr, int x) {
    if (itr == NULL) return NULL;
    else if(x < itr -> data) return find(itr -> left, x);
    else if(x > itr -> data) return find(itr -> right, x);
    else return itr;
  }

  node* findMin_(node* itr) {
    if (itr == NULL) return NULL;
    else if (itr -> left == NULL)  return itr;
    else return findMin_(itr -> left);
  }

  node* findMax(node* itr) {
    if (itr == NULL)  return NULL;
    else if (itr -> right == NULL) return itr;
    else return findMax(itr -> right);
  }
};

int main() {
  BST tree;
  tree.insert(100);
  tree.insert(50);
  tree.insert(20);
  tree.insert(80);
  tree.insert(150);
  tree.insert(120);
  tree.insert(200);
  tree.remove(20);
  tree.remove(120);
  tree.remove(100);

  tree.inOrder();
  tree.preOrder();
  tree.postOrder();
  tree.levelOrder();

  //tree.search(80);
  tree.findMin();
  return 0;
}

File: binarySearchTree.java
public class Main {
  public static void main(String[] args) {
    BinarySearchTree tree = new BinarySearchTree();
    tree.add(10);
    tree.add(5);
    tree.add(2);
    tree.add(8);
    tree.add(15);
    tree.add(12);
    tree.add(20);
    tree.inOrder();
    tree.preOrder();
    tree.postOrder();
    tree.levelOrder();
  }
}




class Queue extends BinarySearchTree {
  class QueueNode {
    Node data;
    QueueNode next;

    public QueueNode(Node node) {
      this.data = node;
      this.next = null;
    }
  }

  public QueueNode head = null;
  public QueueNode tail = null;

  public void enqueue(Node node) {
    QueueNode queueNode = new QueueNode(node);
    if (head == null) {
      head = queueNode;
      tail = head;
    } else {
      tail.next = queueNode;
      tail = queueNode;
    }
  }

  public QueueNode dequeue() {
    QueueNode itr = head;
    if (head == null) return null;
    else if (head == tail) {
      head = null;
      tail = null;
      return itr;
    } else {
      head = head.next;
      return itr;
    }
  }
}




class BinarySearchTree {
  class Node {
    int data;
    Node left;
    Node right;

    public Node(int data) {
      this.data = data;
      this.left = left;
      this.right = right;
    }
  }

  public Node root = null;

  public void add(int data) {
    Node node = root;
    if (node == null) root = new Node(data);
    else searchTree(node, data);
  }

  public void searchTree(Node node, int data) {
    if (node == null) return;
    if (data < node.data) {
      if (node.left == null) node.left = new Node(data);
      else searchTree(node.left, data);
    } else if (data > node.data) {
      if (node.right == null) node.right = new Node(data);
      else searchTree(node.right, data);
    } else System.out.println("Data already in tree");
  }

  public void inOrder() {
    Node node = root;
    System.out.print("In order: ");
    inOrder_(node);
    System.out.println();
  }

  public void inOrder_(Node node) {
    if (node.left != null) inOrder_(node.left);
    System.out.print(node.data + " ");
    if (node.right != null) inOrder_(node.right);
  }

  public void preOrder() {
    Node node = root;
    System.out.print("Pre order: ");
    preOrder_(node);
    System.out.println();
  }

  public void preOrder_(Node node) {
    System.out.print(node.data + " ");
    if (node.left != null) preOrder_(node.left);
    if (node.right != null) preOrder_(node.right);
  }

  public void postOrder() {
    Node node = root;
    System.out.print("Post order: ");
    postOrder_(node);
    System.out.println();
  }

  public void postOrder_(Node node) {
    if (node.left != null) postOrder_(node.left);
    if (node.right != null) postOrder_(node.right);
    System.out.print(node.data + " ");
  }

  public void levelOrder() {
    Queue queue = new Queue();
    Node node = root;
    queue.enqueue(node);
    System.out.print("Level order: ");
    while (queue.head != null) {
      node = queue.dequeue().data;
      System.out.print(node.data + " ");
      if (node.left != null) queue.enqueue(node.left);
      if (node.right != null) queue.enqueue(node.right);
    }
    System.out.println();
  }
}
